### ISSUE TITLE
Vulnerability Detector: don't report vulnerabilities related to inactive kernel packages

### ISSUE THREAD

**Issue by crolopez** [wazuh/wazuh#3635]
When Vulnerability Detector requests the agent packages to WazuhDB, it receives all the installed packages on that machine, including those that are not loaded or enabled.

However, we are not always going to want this. For example, it is a common practice to keep the old kernel images for backup after an upgrade. In these case, all kernel images are considered installed packages by Syscollector, without differentiating which of them was loaded.

Vulnerability Detector should only process or ask for the loaded kernels to avoid false positives.

A possible way to detect the loaded kernel is to check the `OS_UNAME` field from the `AGENTS` table in `GLOBAL.DB`. This column contains the version of the booted kernel, which can allow us to ignore the rest.

``` BASH
sqlite3 /var/ossec/var/db/global.db "SELECT OS_UNAME FROM AGENT WHERE ID = 1;"
Linux |centos73-sources |3.10.0-514.26.2.el7.x86_64 |#1 SMP Tue Jul 4 15:04:05 UTC 2017 |x86_64
```

In the previous output, we can see that the agent has loaded a kernel whose version is `3.10.0-514.26.2.el7.x86_64`.

In the following command output, we can see that this value matches with one of the installed kernels.

``` BASH
rpm -q kernel 
kernel-3.10.0-514.el7.x86_64
kernel-3.10.0-514.26.2.el7.x86_64
```

**Comment by branchnetconsulting** [comment]
This would seriously cut down on the irrelevant noise generated by Vulnerability Detector.  I agree by default we should one way or another exclude all packages that are part of a kernel installed but not loaded on a given agent, from Vulnerability Detector consideration.

**Comment by CODE-REaD** [comment]
Thank you, @crolopez and @branchnetconsulting, having just spent several hours researching why Wazuh was reporting CVE-2019-3846 (and many other vulnerabilities) three times for my CentOS 7 system, I agree!

I finally caught on when I noticed that Wazuh reporting different values in its _data.vulnerability.package.version_ field for each of its otherwise-identical alerts.  I then ran `rpm -qa | grep ^kernel` which confirmed that the kernel releases stored on my system (only one of which is running of course) matched those Wazuh was alerting on.

For now I have run `yum autoremove kernel` to avoid these false positives, but I found nothing in Wazuh's alerts or documentation indicating this requirement.

On further research on this issue I discovered:

- At least with _yum_, the standard is to remove unused packages when an _upgrade_ or _update_ operation is performed.  However, an exception is made (by default) for the following packages:
_ kernel, kernel-bigmem, kernel-enterprise, kernel-smp,  kernel-debug,  kernel-unsupported,  kernel-source, kernel-evel, kernel-PAE, kernel-PAE-debug_.  See _yum.conf(5)_ for details.
- By default, **_yum leaves in place three versions of any of the above packages installed at one time_** (`installonly_limit`).

Bottom Line: as currently coded **_Wazuh will alert on inactive kernel vulnerabilities for any system whose kernel has been updated using yum_**.

**Question:** How does Vulnerability Detector request installed packages?   If it runs a standard system command like `yum` or `rpm` there may be a simple way to tune this part of Wazuh to filter unused packages.

**Comment by branchnetconsulting** [comment]
Yes this really needs to be fixed.  The Wazuh manager already knows at any
point in time what is the precise version of the kernel running on
individual Linux agents, and it seems that the vulnerability-detector
module could be adapted to suppress alerts about kernel-version-locked
packages when the version of the package does not match the version of the
running Linux kernel.

Kevin

**Comment by CODE-REaD** [comment]
@chemamartinez, I note that you removed this issue from To do in Wazuh 3.12.0.  I also note that there is now a Wazuh 4.0.0 project, but do not see this issue listed in its To Do list.  Does that mean a) this issue is resolved in Wazuh 4.0.0, b) this issue has yet to be tested against Wazuh 4.0.0, or c) there is another reason for removing this issue from To do in Wazuh 3.12.0?

I ask because in my case (Linode cloud servers), not even the active kernel is related to a kernel package found on the server ![image](https://user-images.githubusercontent.com/13523087/73855999-747d8280-47fa-11ea-8af1-c0700b4ada38.png).  Because typically three This causes Wazuh to report each matched kernel vulnerability up to three times per linode (due to yum's policy of retaining the three most recent kernel installations; see https://github.com/wazuh/wazuh/issues/3635#issuecomment-572195351).

I am willing to assist in solving this problem if it is still unaddressed.  However I am not familiar with the design or coding of Vulnerability Detector so may need some guidance initially.  Please update me on status if I may assist.  Thank you.

**Comment by chemamartinez** [comment]
Hi @CODE-REaD,

We have moved this issue from the 3.12.0 because it is the next version to be released, and the main effort for it is focused on a big rework for the FIM module. On the other hand, it is not in the 4.0 project because the next major version has not an ETA yet.

It doesn't mean we've dismissed the problem. We are aware of it, and others affecting the vulnerability detector, which is a top priority for us. Is that why I can ensure you we are going to face it in a short/medium-term, along with a review of the entire module to make it more reliable.

I want to thank you for offering to help us solve the problem. As an open-source project, the community is the most important part of it and we are here to help users who want to improve it. Therefore, if you are encouraged to tackle it and need any guidance in doing so, please do not hesitate to ask us.

Regards,
Chema.

**Comment by CODE-REaD** [comment]
@chemamartinez, thank you for your offer of guidance.  I intend to make modifications to `wazuh/src/wazuh_modules/syscollector/syscollector_linux.c` as a "pull request."  I assume I should fork the latest version of this file, i.e., https://github.com/wazuh/wazuh/commit/0ecadb3eb17a20230b8f2c7b5845901c1e613576#diff-a3f78a1923b9cb48894f5a11deb50f86.  Correct?

**Comment by chemamartinez** [comment]
Hi @CODE-REaD,

If you want to submit a pull request with the desired changes. I suggest you fork the last stable version of Wazuh, the `stable` branch in this case. If you need any assistance in the process, don't hesitate to ask us.

Best regards.

**Comment by chemamartinez** [comment]
This issue has been solved at #4996 and will be available in the next release.

Thanks @0xN3utr0n for fixing it.

Regards.
